lib/livebook/notebook/explore/vm_introspection.livemd

```diff
@@ -0,0 +1,268 @@
+# Fun with VM introspection
+
+## Introduction
+
+In this notebook we manually establish connection to a running node,
+and then we try to retrieve and plot some interesting information
+about the system.
+
+## Setup
+
+We are definitely gonna plot some data in this notebook,
+so let's add `:vega_lite` and `:kino` for that.
+
+```elixir
+Mix.install([
+  {:vega_lite, "~> 0.1.0"},
+  {:kino, "~> 0.1.0"}
+])
+```
+
+```elixir
+alias VegaLite, as: Vl
+```
+
+## Connecting to a remote node
+
+This time we will connect to a running node manually instead of
+injecting Livebook there, because we may not want to run `Mix.install/1`
+inside it. This will also give you a better idea of how you can
+inspect a remote node in general.
+
+The first thing we need is an actually running node,
+if you don't have one at hand, feel free to start one with:
+
+`iex --sname cat --cookie meow`
+
+Next, specify the node you want to connect to below
+with the corresponding cookie:
+
+<!-- livebook:{"livebook_object":"cell_input","name":"Node","type":"text","value":"cat@your_hostname"} -->
+
+<!-- livebook:{"livebook_object":"cell_input","name":"Cookie","type":"text","value":"meow"} -->
+
+```elixir
+node =
+  IO.gets("Node: ")
+  |> String.trim()
+  |> String.to_atom()
+
+cookie =
+  IO.gets("Cookie: ")
+  |> String.trim()
+  |> String.to_atom()
+
+Node.set_cookie(node, cookie)
+true = Node.connect(node)
+```
+
+Having successfully connected, let's try spawning a process
+on the remote node!
+
+```elixir
+Node.spawn(node, fn ->
+  IO.inspect(node())
+end)
+```
+
+From the result of `node/1` it's clear that the function was evaluated
+remotely, but note that we still get the standard output back.
+
+## LiveDashboard
+
+[LiveDashboard](https://github.com/phoenixframework/phoenix_live_dashboard)
+is a great tool for getting information and metrics about a running system
+and you can embed it into your Phoenix application very easily. In fact
+even Livebook does that, [check this out](/dashboard).
+
+In the upper right corner you can select the interesting node,
+but at this point you won't see our remote node there!
+
+**Why?**
+
+By default Erlang nodes create a fully connected mesh, meaning that
+each node is directly connected to all other nodes.
+However, the default Livebook runtime is started as a *hidden* node
+for better isolation and consequently its connections are not reflected
+to other nodes. That's the current state:
+
+```
+(Livebook application) <---> (Livebook runtime) <---> (Remote node)
+```
+
+**How?**
+
+So we "are" in `Livebook runtime` and our task is to connect
+`Livebook application` with `Remote node`.
+
+In fact, we already know how to connect to the remote node, we did that earlier.
```

Comment:
Running node directly against the Livebook server sounds a bit crazy. Can we flip it around and instead ask the remote node to connect to the Livebook server? :)

Reply:
I was wondering, but then we need to get the cookie, so it's a bit more complicated than connecting this way round (especially that this way we connect two times pretty much the same way).

Reply:
Ah, good point. Ok, let's keep it as is!